rllib/algorithms/algorithm.py

```diff
@@ -941,6 +943,199 @@ def duration_fn(num_units_done):
+        unit = self.config["evaluation_duration_unit"]
+        eval_cfg = self.config["evaluation_config"]
+        rollout_len = eval_cfg["rollout_fragment_length"]
+        num_envs = eval_cfg["num_envs_per_worker"]
+        auto = self.config["evaluation_duration"] == "auto"
+        duration = (
+            self.config["evaluation_duration"]
+            if not auto
+            else (self.config["evaluation_num_workers"] or 1)
+            * (1 if unit == "episodes" else rollout_len)
+        )
+
+        # Call the `_before_evaluate` hook.
+        self._before_evaluate()
+
+        # Put weights only once into object store and use same object
+        # ref to synch to all workers.
+        self._evaluation_weights_seq_number += 1
+        weights_ref = ray.put(self.workers.local_worker().get_weights())
+        self._sync_filters_if_needed(
+            from_worker=self.workers.local_worker(),
+            workers=self.evaluation_workers,
+            timeout_seconds=eval_cfg.get(
+                "sync_filters_on_rollout_workers_timeout_s"
+            ),
+        )
+
+        if self.config["custom_eval_function"]:
+            raise ValueError(
+                "`custom_eval_function` not supported in combination "
+                "with `evaluation_with_async_requests=True` config setting!"
+            )
+        if (
+            self.evaluation_workers is None
+            and (
+                self.workers.local_worker().input_reader is None
+                or self.config["evaluation_num_workers"] == 0
+            )
+        ):
+            raise ValueError(
+                "Local evaluation OR evaluation without input reader OR evaluation "
+                "with only a local eval worker not supported in combination "
+                "with `evaluation_with_async_requests=True` config setting!"
+            )
+
+        agent_steps_this_iter = 0
+        env_steps_this_iter = 0
+
+        logger.info(f"Evaluating current policy for {duration} {unit}.")
+
+        all_batches = []
+
+        # Default done-function returns True, whenever num episodes
+        # have been completed.
+        if duration_fn is None:
+            def duration_fn(num_units_done):
+                return duration - num_units_done
+
+        def remote_fn(worker, w_ref, w_seq_no):
+            worker.set_weights(weights=w_ref)
+            batch = worker.sample()
+            metrics = worker.get_metrics()
+            return batch, metrics, w_seq_no
+
+        rollout_metrics = []
+
+        # How many episodes have we run (across all eval workers)?
+        num_units_done = 0
+        round_ = 0
+        time_started = time.time()
+        timed_out = True
+
+        while time.time() - time_started < self.config["evaluation_sample_timeout_s"]:
+            units_left_to_do = duration_fn(num_units_done)
+            if units_left_to_do <= 0:
+                timed_out = False
+                break
+
+            round_ += 1
```

Comment:
rename to _round? don't know if style guide suggests training _ for any type of variables.

Reply:
done